src/backend/functions/work-packages-edit.ts

```diff
@@ -0,0 +1,251 @@
+      return element.detail;
+    }),
+    expectedActivities,
+    crId,
+    userId,
+    wbsElementId,
+    'expected activity'
+  );
+  const deliverablesChangeJson = createListChangesJson(
+    originalWorkPackage.deliverables.map((element) => {
+      return element.detail;
+    }),
+    deliverables,
+    crId,
+    userId,
+    wbsElementId,
+    'deliverable'
+  );
+
+  // add to changes if not undefined
+  if (nameChangeJson !== undefined) {
+    changes.push(nameChangeJson);
+  }
+  if (projectIdChangeJson !== undefined) {
+    changes.push(projectIdChangeJson);
+  }
+  if (startDateChangeJson !== undefined) {
+    changes.push(startDateChangeJson);
+  }
+  if (durationChangeJson !== undefined) {
+    changes.push(durationChangeJson);
+  }
+
+  changes = changes
+    .concat(dependenciesChangeJson)
+    .concat(expectedActivitiesChangeJson)
+    .concat(deliverablesChangeJson);
+
+  // update the work package with the input fields
+  const updatedWorkPackage = await prisma.work_Package.update({
+    where: {
+      wbsElementId
+    },
+    data: {
+      projectId,
+      startDate,
+      duration,
+      wbsElement: {
+        update: {
+          name
+        }
+      },
+      dependencies: {
+        connect: wbsElementIds.map((ele: any) => {
+          return { wbsElementId: ele };
+        })
+      },
+      expectedActivities: {
+        create: expectedActivities.map((ele: any) => {
+          return { detail: ele };
+        })
+      },
+      deliverables: {
+        create: deliverables.map((ele: any) => {
+          return { detail: ele };
+        })
+      }
+    }
+  });
+
+  // create the changes in prisma
+  await prisma.change.createMany({
+    data: changes
+  });
+
+  // return the updated work package
+  return buildSuccessResponse(updatedWorkPackage);
+};
+
+// create a change json if the old and new value are different, otherwise return undefined
+const createChangeJsonNonList = (
+  nameOfField: string,
+  oldValue: any,
+  newValue: any,
+  crId: number,
+  implementerId: number,
+  wbsElementId: number
+) => {
+  if (oldValue !== newValue) {
+    return {
+      changeRequestId: crId,
+      dateImplemented: new Date(),
+      implementerId,
+      wbsElementId,
+      detail: `Changed ${nameOfField} from ${oldValue} to ${newValue}`
+    };
+  }
+  return undefined;
+};
+
+// create a change json list for a given list (dependencies). Only works if the elements can only be added or deleted
+const createListChangesJson = <T>(
+  oldArray: T[],
+  newArray: T[],
+  crId: number,
+  implementerId: number,
+  wbsElementId: number,
+  nameOfField: string
+) => {
+  const seenOld = new Set<T>();
+  const seenNew = new Set<T>();
+  oldArray.forEach((element: T) => {
+    seenOld.add(element);
+  });
+
+  newArray.forEach((element: T) => {
+    seenNew.add(element);
+  });
+
+  const changes: { element: T; type: string }[] = [];
+
+  oldArray.forEach((element: T) => {
+    if (seenNew.has(element)) {
```

Comment:
does this mean: for each element in the original array, if it also exists in the new array, then create a change for removing it? Cuz that doesn't sound right to me. I'm a tad confused as to how the code works

Reply:
yeah this was a bug it's actually if !seenNew.has